docs/src/modules/ROOT/pages/enterprise-edition/enterprise-edition.adoc

```diff
@@ -716,5 +717,215 @@ the host is likely to hang or freeze,
+}
+----
+
+The Java Virtual Machine is free to (and often does) create different instances of functionality equivalent lambdas.
+This severely limits the effectiveness of node sharing, since the only way to know two lambdas are equal is to compare their references.
+
+When automatic node sharing is used, the `ConstraintProvider` class is transformed so all lambdas are accessed via a static final field.
+So a class that looks like this:
+
+[source,java,options="nowrap"]
+----
+public class MyConstraintProvider implements ConstraintProvider {
+
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+            a(constraintFactory),
+            b(constraintFactory)
+        };
+    }
+
+    Constraint a(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter(shift -> shift.getEmployee().getName().equals("Ann"))
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("a");
+    }
 
+    Constraint b(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter(shift -> shift.getEmployee().getName().equals("Ann"))
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("b");
+    }
+}
+----
+
+is transformed to a class that looks like this:
+
+[source,java,options="nowrap"]
+----
+public class MyConstraintProvider implements ConstraintProvider {
+    private static final Predicate<Shift> $predicate1 = shift -> shift.getEmployee().getName().equals("Ann");
+
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+            a(constraintFactory),
+            b(constraintFactory)
+        };
+    }
+
+    Constraint a(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter($predicate1)
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("a");
+    }
+
+    Constraint b(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter($predicate1)
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("b");
+    }
+}
+----
+
+IMPORTANT: This transformation prevents breakpoints from being placed inside the ConstraintProvider.
+
+This allows building blocks to share the same nodes without needing the `ConstraintProvider` to be written in an awkward way.
+
+==== Configuration
+
+[tabs]
+======
+Plain Java::
+* Add `ai.timefold.solver.enterprise:timefold-solver-enterprise-asm` to your `pom.xml`:
++
+[source,xml,options="nowrap"]
+----
+<!-- ... -->
+<dependencies>
+  <dependency>
+    <groupId>ai.timefold.solver.enterprise</groupId>
+    <artifactId>timefold-solver-enterprise-asm</artifactId>
+  </dependency>
+</dependencies>
```

Comment:
This should not be necessary. This is boilerplate, we need Enterprise to be easy. Make `enterprise-core` depend on ASM by default.

Reply:
I was proposing something less drastic, I think - Make `asm` JAR a dependency of the `enterprise-core` JAR. If they can add it as a dependency anyway, there's no reason why we shouldn't do it for them.

But maybe we should first decide if we want this feature enabled by default. If not, then obviously we should not add this default dependency.